api/rosetta/validator.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-playground/validator/v10"
+
+	"github.com/optakt/flow-dps/rosetta/identifier"
+)
+
+var (
+	// Special case in the validator library that shouldn't really happen with correct usage.
+	errInvalidValidation = errors.New("invalid validation input")
+
+	errBlockLength   = errors.New(blockLength)
+	errAddressLength = errors.New(addressLength)
+	errTxLength      = errors.New(txLength)
+)
+
+// Field names needed by the validator library, not needed or used by our code at the moment.
+const (
+	blockHashField   = "block_hash"
+	blockchainField  = "blockchain"
+	networkField     = "network"
+	addressField     = "address"
+	txField          = "transaction_id"
+	currencyField    = "currency"
+	symbolField      = "symbol"
+	transactionField = "transaction"
+	signaturesField  = "signatures"
+)
+
+type Validator struct {
```

Comment:
This could be unexported, but I was juggling multiple entities and verbs and terms validator/validate was all over the place and it started to get confusing what is referenced. This felt like it provided a clear distinction what is what.

I'm open to suggestions on a different name 😄

I think this whole validation logic could be offloaded to a separate package, perhaps `api/rosetta/validator` or something like that.

Reply:
We already have a `api/rosetta/validator` package. Why isn't this logic included in it? 🤔

Reply:
There is `rosetta/vlidator` (vs `api/rosetta/validator`). That one deals with the Rosetta business logic, while this one deals with the inputs for the Rosetta API.

Theoretically they could be merged since they serve the same purpose, but I also feel they kind of deal with different things.

Reply:
I agree that it could be in the same package.

Reply:
Ah right my bad, but yeah I still think indeed that they should be together, having two separate rosetta validators would be confusing IMO, and there's no downside to merging them AFAIK.